pkg/app/api/grpcapi/web_api.go

```diff
@@ -599,6 +602,85 @@ func (a *WebAPI) validatePipedBelongsToProject(ctx context.Context, pipedID, pro
 	return nil
 }
 
+func (a *WebAPI) ListUnregisteredApplications(ctx context.Context, _ *webservice.ListUnregisteredApplicationsRequest) (*webservice.ListUnregisteredApplicationsResponse, error) {
+	claims, err := rpcauth.ExtractClaims(ctx)
+	if err != nil {
+		a.logger.Error("failed to authenticate the current user", zap.Error(err))
+		return nil, err
+	}
+
+	// Collect all apps that belong to the project.
+	key := makeUnregisteredAppsCacheKey(claims.Role.ProjectId)
+	c := rediscache.NewHashCache(a.redis, key)
+	// pipedToApps assumes to be a map["piped-id"][]*model.ApplicationInfo
+	pipedToApps, err := c.GetAll()
+	if errors.Is(err, cache.ErrNotFound) {
+		return &webservice.ListUnregisteredApplicationsResponse{}, nil
+	}
+	if err != nil {
+		a.logger.Error("failed to get unregistered apps", zap.Error(err))
+		return nil, status.Error(codes.Internal, "Failed to get unregistered apps")
+	}
+
+	// Integrate all apps cached for each Piped.
+	allApps := make([]*model.ApplicationInfo, 0)
+	for _, as := range pipedToApps {
+		apps, ok := as.([]*model.ApplicationInfo)
+		if !ok {
+			return nil, status.Error(codes.Internal, "Unexpected data cached")
+		}
+		allApps = append(allApps, apps...)
+	}
+
+	return &webservice.ListUnregisteredApplicationsResponse{
+		Repos: groupAppsByRepo(allApps),
+	}, nil
+}
+
+func groupAppsByRepo(apps []*model.ApplicationInfo) []*webservice.ListUnregisteredApplicationsResponse_Repo {
+	if len(apps) == 0 {
+		return []*webservice.ListUnregisteredApplicationsResponse_Repo{}
+	}
+	if len(apps) == 1 {
+		return []*webservice.ListUnregisteredApplicationsResponse_Repo{
+			{Id: apps[0].RepoId, Apps: apps},
+		}
+	}
+
+	// Make a map from repo-id to apps.
+	repoToApps := make(map[string][]*model.ApplicationInfo)
+	for _, app := range apps {
+		if _, ok := repoToApps[app.RepoId]; !ok {
+			repoToApps[app.RepoId] = []*model.ApplicationInfo{}
+		}
+		repoToApps[app.RepoId] = append(repoToApps[app.RepoId], app)
+	}
+
+	// Tidy apps.
+	repos := make([]*webservice.ListUnregisteredApplicationsResponse_Repo, 0, len(repoToApps))
+	for repoID, as := range repoToApps {
+		// Eliminate duplicated apps
+		tidiedApps := make([]*model.ApplicationInfo, 0, len(as))
+		gitPaths := make(map[string]struct{})
+		for _, app := range as {
+			if _, ok := gitPaths[app.GetPath()]; ok {
+				continue
+			}
+			gitPaths[app.GetPath()] = struct{}{}
+			tidiedApps = append(tidiedApps, app)
+		}
+
+		sort.Slice(tidiedApps, func(i, j int) bool {
+			return tidiedApps[i].GetPath() < tidiedApps[j].GetPath()
+		})
+		repos = append(repos, &webservice.ListUnregisteredApplicationsResponse_Repo{
+			Id:   repoID,
+			Apps: tidiedApps,
+		})
+	}
+	return repos
```

Comment:
There are some concerns around the way we are grouping applications by repo-id and eliminating duplicates by GitPath.

Since we don't have any guide or restriction on the repository ID across Piped currently, so users can use the same repo ID in different Pipeds for different repositories.

For example:
- piped-dev has a `manifests` repo-id which is referring to `manifests-dev` repository
- piped-prod also has a `manifests` repo-id which is referring to `manifests-prod` repository

That will result in a not-correct list.

So I think we can return the raw data from the cache that includes PipedID in the response.

Reply:
For sure. Obviously repo-id isn't a project-wide concept. Apparently It's enough to give back directly grouped by Piped.

Reply:
The response structure depends on how to show the list on the adding form. Of course it will be able to change, but tell me a little bit more about your thoughts.

I'm guessing the unregistered list on the web is going to be kind of like:
- piped-1
  - app-1 (repo: "repo-1", path: "path/to", cfgFilename: ".pipe.yaml")
  - app-2 (repo: "repo-2", path: "path/to", cfgFilename: ".pipe.yaml")
- piped-2
  - ...

Therefore, the response of ListUnregisteredApplications would be like:

```proto
message ListUnregisteredApplicationsResponse {
    message Piped {
        string id = 1;
        repeated model.ApplicationInfo apps = 2;
    }
    repeated Piped pipeds = 1;
}
```

But if it's enough to show a flat list like:
- app-1 (piped: "piped-1", repo: "repo-1", path: "path/to", cfgFilename: ".pipe.yaml")
- app-2 (piped: "piped-2", repo: "repo-1", path: "path/to", cfgFilename: ".pipe.yaml")

just add a field called PipedId to ApplicationInfo and return a list of ApplicationInfo.